File: ChessLab/Variants/Chess960Position.swift
import Foundation

/// Les 960 positions de départ du Chess960, par leur numéro de Scharnagl.
///
/// L'algorithme est le PORT FIDÈLE de `chess.BaseBoard._set_chess960_pos` de
/// python-chess (lui-même d'après russellcottrell.com/Chess/Chess960.htm) —
/// et le test `Chess960RulesTests` compare les 960 FEN produits à ceux de
/// python-chess, généré par `tools/opening-generator/gen_chess960_fixtures.py`.
/// La numérotation est donc INTERCHANGEABLE avec Lichess et les moteurs :
/// la position 518 est la partie classique.
enum Chess960Position {

    /// Rangée de base (blanche, a→h) du numéro `number` ∈ 0...959.
    static func backRank(number: Int) -> [Character]? {
        guard (0...959).contains(number) else { return nil }

        var n = number
        let bw = n % 4; n /= 4          // fou de cases claires : b, d, f, h
        let bb = n % 4; n /= 4          // fou de cases sombres : a, c, e, g
        let q = n % 6; n /= 6           // dame : q-ième case libre

        // Décomposition N5N : n ∈ 0...9 désigne la paire de cases des
        // cavaliers parmi les 5 restantes — même boucle que python-chess.
        var n1 = 0, n2 = 0
        for candidate in 0..<4 {
            n1 = candidate
            n2 = n + (3 - candidate) * (4 - candidate) / 2 - 5
            if n1 < n2, (1...4).contains(n2) { break }
        }

        var rank = [Character?](repeating: nil, count: 8)
        let bwFile = bw * 2 + 1
        let bbFile = bb * 2
        rank[bwFile] = "B"
        rank[bbFile] = "B"

        var qFile = q
        if min(bwFile, bbFile) <= qFile { qFile += 1 }
        if max(bwFile, bbFile) <= qFile { qFile += 1 }
        rank[qFile] = "Q"

        var counterN1 = n1, counterN2 = n2
        for file in 0..<8 where rank[file] == nil {
            if counterN1 == 0 || counterN2 == 0 {
                rank[file] = "N"
            }
            counterN1 -= 1
            counterN2 -= 1
        }

        // Les trois cases restantes, de gauche à droite : tour, roi, tour.
        for piece in "RKR" {
            guard let file = rank.firstIndex(of: nil) else { return nil }
            rank[file] = piece
        }
        return rank.compactMap(\.self)
    }

    /// La rangée `rank` (8 lettres, a→h) respecte-t-elle les règles du
    /// Chess960 ? Trois conditions, les mêmes que l'algorithme de Scharnagl
    /// impose par construction — nécessaires ici parce qu'un agencement
    /// composé À LA MAIN peut les violer :
    /// - le ROI est strictement ENTRE les deux tours (sans quoi le roque tel
    ///   que cette app le joue — « le roi prend sa tour » — perd son sens : on
    ///   ne saurait plus quelle tour est « petit côté ») ;
    /// - les deux FOUS sont sur des cases de couleurs DIFFÉRENTES (règle FIDE
    ///   du Chess960, pas une contrainte technique de cette app) ;
    /// - le jeu de pièces exact (2 tours, 2 cavaliers, 2 fous, 1 dame, 1 roi)
    ///   — automatiquement vrai si `rank` est un RÉARRANGEMENT d'une rangée
    ///   déjà légale (l'éditeur ne fait que permuter), mais vérifié quand
    ///   même : la fonction doit rester correcte pour tout appelant.
    static func isLegalBackRank(_ rank: [Character]) -> Bool {
        guard rank.count == 8, rank.sorted() == Array("BBKNNQRR") else { return false }
        guard let kingFile = rank.firstIndex(of: "K") else { return false }
        let rookFiles = rank.indices.filter { rank[$0] == "R" }
        guard rookFiles.count == 2, rookFiles[0] < kingFile, kingFile < rookFiles[1] else { return false }
        let bishopFiles = rank.indices.filter { rank[$0] == "B" }
        guard bishopFiles.count == 2 else { return false }
        return bishopFiles[0].isMultiple(of: 2) != bishopFiles[1].isMultiple(of: 2)
    }

    /// Numéro de Scharnagl (0...959) dont la rangée de base est EXACTEMENT
    /// `rank` — `nil` si elle n'est pas légale. Balayage linéaire : 960
    /// positions, coût négligeable, et ça évite d'inverser l'algorithme de
    /// génération pour un besoin qui ne se produit qu'au réglage d'une
    /// partie, jamais en boucle chaude.
    static func number(forBackRank rank: [Character]) -> Int? {
        guard isLegalBackRank(rank) else { return nil }
        return (0...959).first { backRank(number: $0) == rank }
    }

    /// FEN Shredder complet de la position `number` : droits de roque en
    /// lettres de colonnes (ex. `HFhf`), le dialecte que parle Stockfish
    /// sous `UCI_Chess960`.
    static func startingFEN(number: Int) -> String? {
        guard let rank = backRank(number: number) else { return nil }
        let white = String(rank)
        let black = white.lowercased()
        let rookFiles = rank.indices.filter { rank[$0] == "R" }
            .map { Character(UnicodeScalar(UInt8(65 + $0))) }
        // Ordre Shredder : colonnes décroissantes côté blanc (HAha pour la 518).
        let rights = String(rookFiles.sorted(by: >))
        return "\(black)/pppppppp/8/8/8/8/PPPPPPPP/\(white) w \(rights)\(rights.lowercased()) - 0 1"
    }
}

File: android/app/src/main/java/com/chesslab/variants/Chess960Position.kt
package com.chesslab.variants

/**
 * Les 960 positions de départ du Chess960, par leur numéro de Scharnagl.
 * Pendant de `Chess960Position.swift`.
 *
 * L'algorithme est le port fidèle de `chess.BaseBoard._set_chess960_pos` de
 * python-chess, et le test compare les 960 FEN produites à celles que
 * python-chess a écrites (`ChessLabTests/Fixtures_chess960_starts.json`, le
 * MÊME fichier que la suite iOS). La numérotation est donc INTERCHANGEABLE
 * avec Lichess et les moteurs : la position 518 est la partie classique.
 */
object Chess960Position {

    /** La position classique : sa numérotation est celle de tout le monde. */
    const val classic = 518

    val range = 0..959

    /** Rangée de base (blanche, a→h) du numéro [number], ou `null` hors bornes. */
    fun backRank(number: Int): CharArray? {
        if (number !in range) return null

        var n = number
        val bw = n % 4; n /= 4          // fou de cases claires : b, d, f, h
        val bb = n % 4; n /= 4          // fou de cases sombres : a, c, e, g
        val q = n % 6; n /= 6           // dame : q-ième case libre

        // Décomposition N5N : n ∈ 0…9 désigne la paire de cases des cavaliers
        // parmi les 5 restantes — même boucle que python-chess.
        var n1 = 0
        var n2 = 0
        for (candidate in 0 until 4) {
            n1 = candidate
            n2 = n + (3 - candidate) * (4 - candidate) / 2 - 5
            if (n1 < n2 && n2 in 1..4) break
        }

        val rank = arrayOfNulls<Char>(8)
        val bwFile = bw * 2 + 1
        val bbFile = bb * 2
        rank[bwFile] = 'B'
        rank[bbFile] = 'B'

        var qFile = q
        if (minOf(bwFile, bbFile) <= qFile) qFile++
        if (maxOf(bwFile, bbFile) <= qFile) qFile++
        rank[qFile] = 'Q'

        var counterN1 = n1
        var counterN2 = n2
        for (file in 0 until 8) {
            if (rank[file] != null) continue
            if (counterN1 == 0 || counterN2 == 0) rank[file] = 'N'
            counterN1--
            counterN2--
        }

        // Les trois cases restantes, de gauche à droite : tour, roi, tour.
        for (piece in "RKR") {
            val file = rank.indexOfFirst { it == null }
            if (file < 0) return null
            rank[file] = piece
        }
        return CharArray(8) { rank[it]!! }
    }

    /**
     * La rangée respecte-t-elle les règles du Chess960 ? Trois conditions, que
     * l'algorithme de Scharnagl impose par construction — nécessaires ici
     * parce qu'un agencement composé À LA MAIN peut les violer :
     * - le ROI est strictement ENTRE les deux tours (sans quoi le roque tel que
     *   cette app le joue perd son sens : on ne saurait plus quelle tour est
     *   « petit côté ») ;
     * - les deux FOUS sont sur des cases de couleurs DIFFÉRENTES (règle FIDE) ;
     * - le jeu de pièces exact.
     */
    fun isLegalBackRank(rank: CharArray): Boolean {
        if (rank.size != 8) return false
        if (String(rank.clone().also { it.sort() }) != "BBKNNQRR") return false
        val king = rank.indexOf('K')
        if (king < 0) return false
        val rooks = rank.indices.filter { rank[it] == 'R' }
        if (rooks.size != 2 || rooks[0] >= king || king >= rooks[1]) return false
        val bishops = rank.indices.filter { rank[it] == 'B' }
        if (bishops.size != 2) return false
        return (bishops[0] % 2 == 0) != (bishops[1] % 2 == 0)
    }

    /**
     * Le numéro dont la rangée de base est EXACTEMENT [rank], ou `null`.
     * Balayage linéaire : 960 positions, coût négligeable, et ça évite
     * d'inverser l'algorithme pour un besoin qui ne se produit qu'au réglage
     * d'une partie, jamais en boucle chaude.
     */
    fun number(rank: CharArray): Int? {
        if (!isLegalBackRank(rank)) return null
        return range.firstOrNull { backRank(it)?.concatToString() == rank.concatToString() }
    }

    /**
     * FEN Shredder complète de la position : droits de roque en LETTRES DE
     * COLONNES (« HFhf »), le dialecte que parle le moteur sous
     * `UCI_Chess960` — la seule forme non ambiguë quand les tours ne sont pas
     * en a1 et h1.
     */
    fun startingFen(number: Int): String? {
        val rank = backRank(number) ?: return null
        val white = rank.concatToString()
        val black = white.lowercase()
        val rights = rank.indices.filter { rank[it] == 'R' }
            .map { ('A' + it) }
            .sortedDescending()
            .joinToString("")
        return "$black/pppppppp/8/8/8/8/PPPPPPPP/$white w $rights${rights.lowercase()} - 0 1"
    }
}
